Add support for running SYSCLK from PLL1, via PLL2.

attie-argentum opened the pull request:
For parts like the STM32F1 Connectivity Line (STM32F105xx, STM32F107xx), it is occasionally necessary to source SYSCLK via PLL2. This patch will add this support.

- Add: `UTILS_GetPLL2OutputFrequency()` to calculate the output frequency of PLL2
- Add: `LL_PLL_ConfigSystemClock_PLL2()` to configure the system clock as sourced from HSE, via PLL2 and PLL1.
- Add: Miscellaneous support definitions.

This PR relates to zephyrproject-rtos/zephyr#21224.

Ping @erwango

attie-argentum commented:
Hi @ALABSTM,

Is there any progress on this PR?

ALABSTM commented:
Hi @attie-argentum,

Please allow me first to thank you for your proposal. I just forwarded your report to our development teams. I will get back to you as soon as they provide me with their feedback.

With regards,

ALABSTM commented:
Hi @attie-argentum,

I hope you are doing well. According to our development team sourcing the SYSCLK via PLL2 is **already supported** by our firmware. Indeed:
* The function `LL_PLL_ConfigSystemClock_PLL2()` you added into the `stm32f1xx_ll_utils.c` file is almost the same as the function `SystemClock_Config()` present in the `main.c` file of the **LL template project** and can also be generated using the **Cube MX** tool.
https://github.com/STMicroelectronics/STM32CubeF1/blob/276c4ab9537e83b9897badb68e02d9c4c46c5c22/Projects/STM3210C_EVAL/Templates_LL/Src/main.c#L59-L120

* The definitions you added into the `stm32f1xx_ll_rcc.h` file already exist.
https://github.com/STMicroelectronics/STM32CubeF1/blob/276c4ab9537e83b9897badb68e02d9c4c46c5c22/Drivers/STM32F1xx_HAL_Driver/Inc/stm32f1xx_ll_rcc.h#L166-L190

I hope this addresses your need. Do not hesitate if you would like to share you point-of-view and thank you again for your proposal.

With regards,

attie-argentum committed:
Add support for running SYSCLK from PLL1, via PLL2.

For parts like the STM32F1 Connectivity Line (STM32F105xx, STM32F107xx),
it is occasionally necessary to source SYSCLK via PLL2. This patch will
add this support.

- Add: `UTILS_GetPLL2OutputFrequency()` to calculate the output frequency of PLL2
- Add: `LL_PLL_ConfigSystemClock_PLL2()` to configure the system clock as sourced from HSE, via PLL2 and PLL1.
- Add: Miscellaneous support definitions.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum commented:
Hello @ALABSTM,

Thanks very much for your feedback.

I have removed the duplicate definitions - my apologies for not spotting these, thank you for pointing them out.

While I appreciate that the example `SystemClock_Config()` you linked to demonstrates that it is possible to operate via PLL2, I don't think it's quite appropriate for use in other / larger projects (like Zephyr, which is the goal here). We'd need to reproduce this code, and perform all of the configuration accordingly.

The `LL_PLL_ConfigSystemClock_PLL2()` function I've proposed complements the features of `LL_PLL_ConfigSystemClock_HSE()`, but with the inclusion of PLL2 in the clock path.

It would be used like this: [clock_stm32_ll_common.c:403-429](https://github.com/argentum-systems/zephyr/blob/da99de64bcf9cbb3dc8a2e1da2e0a4fb9f45258b/drivers/clock_control/clock_stm32_ll_common.c#L403-L429)

In a previous iteration of my [STM32F105 PR to Zephyr](https://github.com/zephyrproject-rtos/zephyr/pull/21224), I'd implemented most of what I have now relocated to this repo, as it didn't really fit in their code base.

Do you think that given the intended use-case this might be more acceptable / appropriate?

Regards,
Attie

ALABSTM commented:
Hi @attie-argentum,

I hope you are fine. Sorry for the delay it may take sometimes to answer.

I discussed further your change request with our development team and they agreed with the proposal. The way they will implement it may differ from yours in some aspects. I will keep you informed shall there be anything new.

Thank you again for this contribution and thank you in advance for your patience.

With regards,

ALABSTM commented:
ST Internal Reference: 102027

attie-argentum commented:
Hi @ALABSTM - no problem, thanks for letting me know!

ALABSTM commented:
Hi @attie-argentum,

I hope you are fine. Good news! Your proposal has been **accepted** and implemented. The updated will be available in the frame of the next STM32CubeF1 release hopefully. In the meanwhile, your pull-request will be merged to make this enhancement available.

Please note that we slightly changed the internal implementation of the `LL_PLL_ConfigSystemClock_PLL2()`: as the body is very similar to that of `LL_PLL_ConfigSystemClock_HSE()`, we managed to "factorize" some code. However, and most important, we kept the **same prototype**.

Thank you again for your contribution.

With regards,

attie-argentum commented:
Fantastic, thank you!
